pkg/toolsets/toolsets.go

```diff
@@ -58,13 +58,27 @@ func (t *Toolset) SetReadOnly() {
 
 func (t *Toolset) AddWriteTools(tools ...server.ServerTool) *Toolset {
 	// Silently ignore if the toolset is read-only to avoid any breach of that contract
+	for _, tool := range tools {
+		if tool.Tool.Annotations.ReadOnlyHint {
+			panic(fmt.Sprintf("tool (%s) is incorrectly annotated as read-only", tool.Tool.Name))
+		}
+	}
```

Comment:
I think this should probably be tested, but if you're open to it I'd like to flip this and see if we can use the `ReadOnlyHint` declaratively to build the toolsets?